modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/Inbox.java

```diff
@@ -475,6 +482,7 @@ enum State {
         }
 
         private final PriorityQueue<Batch<RowT>> batches = new PriorityQueue<>(IO_BATCH_CNT);
+        private final AtomicReference<SharedState> sharedStateHolder = new AtomicReference<>();
```

Comment:
Seems, it just used to exchange value in a single line of code without implying any atomicity requirement.
Thus it can be replace with plain field, which can be explicitly nullified in a code with the comment like "we won't send shared state more than once until next rewind".

Reply:
fixed